htmd/smallmol/tools/clustering.py

```diff
@@ -0,0 +1,423 @@
+        match = m.GetSubstructMatch(mcs_mol)
+        sel_str = convertToString(match)
+
+        atoms_mcs = sm.get('idx', 'idx {}'.format(sel_str))
+        atoms_no_mcs = sm.get('idx', 'idx {}'.format(sel_str),  invert=True)
+
+        atoms_mcs_list.append(atoms_mcs.tolist())
+        atoms_no_mcs_list.append(atoms_no_mcs.tolist())
+
+    return mcs_mol, atoms_mcs_list, atoms_no_mcs_list
+
+def cluster(smallmol_list, method, distThresholds=0.2, returnDetails=True, removeHs=True ):
+    """
+    Rreturn the SmallMol objects grouped in the cluster. It can also return the details of the clusters computed.
+
+    Parameters
+    ----------
+    smallmol_list: list
+        The list of htmd.smallmol.smallmol.SmallMol objects
+    method: str
+        The cluster methods. Can be  ['maccs', 'pathFingerprints', 'atomsFingerprints', 'torsionsFingerprints',
+        'circularFingerprints', 'shape', 'mcs']
+    distThresholds: float
+        The disance cutoff for the clusters
+        Default: 0.2
+    returnDetails: bool
+        If True, the cluster details are also returned
+        Default: True
+    removeHs: bool
+        If True, the hydrogens are not considered
+        Default: True
+
+    Returns
+    -------
+    clusters: list
+        List of lists, That contains the SmallMol objects grouped based on the cluster belongings
+    details: list
+        A list with all the cluster details
+    """
+
+
+    from sklearn.cluster import DBSCAN
+
+    import sys
+    this_module = sys.modules[__name__]
+
+    _methods = ['maccs', 'pathFingerprints', 'atomsFingerprints', 'torsionsFingerprints',
+                'circularFingerprints', 'shape', 'mcs']
+
+    if method not in _methods:
+        raise ValueError('The method provided {} does not exists. The ones available are the following: {}'.format(method, _methods))
+
+    smallmol_list = np.array([sm.copy() for sm in smallmol_list])
+
+    if removeHs:
+        tmp_smallmol_list = []
+        for sm in smallmol_list:
+            B = Builder(sm)
+            B.removeHydrogens()
+            sm = B.getSmallMol()
+            tmp_smallmol_list.append(sm)
+            #sm._removeAtoms(sm.get('element H', 'idx'))
+
+        smallmol_list = np.array(tmp_smallmol_list)
+
+    #rdkitMols_list = [sm.toRdkitMol(includeConformer=True) for sm in smallmol_list]
+    rdkitMols_list = []
+    wrong = []
+    for n, sm in enumerate(smallmol_list):
+        try:
+            rdkitMols_list.append(sm.toRdkitMol(includeConformer=True))
+        except:
+            wrong.append(n)
+    print('{} problematic molecules. Indexes: {}'.format(len(wrong), wrong))
+
+    clustmethod = getattr(this_module, '_{}Clustering'.format(method))
+
+    if method not in ['shape', 'mcs']:
+        matrix = clustmethod(rdkitMols_list)
+
+    else:
+        aprun = ParallelExecutor(n_jobs=-1)  # _config['ncpus'])
+        matrix = aprun(total=len(rdkitMols_list), desc='{} Distance'.format(method)) \
```

Comment:
You are parallelizing over parallelizations, since your methods also parallelize! Are you really sure you want to do this?